File: leetcode/medium/15-3Sum.py
#!/usr/bin/python3

class Solution:
  # Two pointer solution O(n^3)
  def threeSum(self, nums: list) -> list:
    res = []
    nums = sorted(nums)
    n = len(nums)
    for i in range(n):
      if i != 0 and nums[i] == nums[i - 1]:
        continue
      j = i + 1
      k = n - 1
      while j < k:
        if nums[i] + nums[j] + nums[k] == 0:
          res.append([nums[i], nums[j], nums[k]])
          j += 1
          while j < k and nums[j] == nums[j - 1]:
            j += 1
        elif nums[i] + nums[j] + nums[k] < 0:
          j += 1
        else:
          k -= 1
    return res

File: neetcode150/medium/cpp/15-3Sum.cpp
#include <vector>

using namespace std;

class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> res;
        int curr;
        sort(nums.begin(), nums.end());
        for (int i = 0; i < nums.size()-2; i++){
            if (i > 0 && nums[i] == nums[i - 1]){
                continue;
            }
            int l = i+1;
            int r = nums.size()-1;
            while (l < r) {
                curr = nums[i] + nums[l] + nums[r];
                if (curr == 0){
                    res.push_back({nums[i], nums[l], nums[r]});
                    while (l < r && nums[r] == nums[r-1]) {
                        --r;
                    }
                    while (l < r && nums[l] == nums[l+1]) {
                        ++l;
                    }
                    ++l;
                    --r;
                } else if (curr > 0) {
                    --r;
                } else {
                    ++l;
                }
            } 
        }
        return res;
    }
};